libfreerdp/core/surface.c

```diff
@@ -28,39 +28,74 @@
 
 #define TAG FREERDP_TAG("core.surface")
 
-static BOOL update_recv_surfcmd_surface_bits(rdpUpdate* update, wStream* s, UINT32* length)
+static BOOL update_recv_surfcmd_bitmap_header_ex(wStream* s, TS_COMPRESSED_BITMAP_HEADER_EX* header)
+{
+	if (!s || !header)
+		return FALSE;
+
+	if (Stream_GetRemainingLength(s) < 24)
+		return FALSE;
+
+	Stream_Read_UINT32(s, header->highUniqueId);
+	Stream_Read_UINT32(s, header->lowUniqueId);
+	Stream_Read_UINT64(s, header->tmMilliseconds);
+	Stream_Read_UINT64(s, header->tmSeconds);
+	return TRUE;
+}
+
+static BOOL update_recv_surfcmd_bitmap_ex(wStream* s, TS_BITMAP_DATA_EX* bmp)
 {
 	size_t pos;
-	SURFACE_BITS_COMMAND* cmd = &update->surface_bits_command;
+	if (!s || !bmp)
+		return FALSE;
 
-	if (Stream_GetRemainingLength(s) < 20)
+	if (Stream_GetRemainingLength(s) < 12)
 		return FALSE;
 
-	Stream_Read_UINT16(s, cmd->destLeft);
-	Stream_Read_UINT16(s, cmd->destTop);
-	Stream_Read_UINT16(s, cmd->destRight);
-	Stream_Read_UINT16(s, cmd->destBottom);
-	Stream_Read_UINT8(s, cmd->bpp);
+	Stream_Read_UINT8(s, bmp->bpp);
+	Stream_Read_UINT8(s, bmp->flags);
+	Stream_Seek(s, 1); /* reserved */
+	Stream_Read_UINT8(s, bmp->codecID);
+	Stream_Read_UINT16(s, bmp->width);
+	Stream_Read_UINT16(s, bmp->height);
+	Stream_Read_UINT32(s, bmp->bitmapDataLength);
 
-	if ((cmd->bpp < 1) || (cmd->bpp > 32))
+	if ((bmp->bpp < 1) || (bmp->bpp > 32))
 	{
-		WLog_ERR(TAG, "invalid bpp value %"PRIu32"", cmd->bpp);
+		WLog_ERR(TAG, "invalid bpp value %"PRIu32"", bmp->bpp);
 		return FALSE;
 	}
 
-	Stream_Seek(s, 2); /* reserved1, reserved2 */
-	Stream_Read_UINT8(s, cmd->codecID);
-	Stream_Read_UINT16(s, cmd->width);
-	Stream_Read_UINT16(s, cmd->height);
-	Stream_Read_UINT32(s, cmd->bitmapDataLength);
+	memset(&bmp->exBitmapDataHeader, 0, sizeof(TS_COMPRESSED_BITMAP_HEADER_EX));
+	if (bmp->flags & EX_COMPRESSED_BITMAP_HEADER_PRESENT)
+	{
+		if (!update_recv_surfcmd_bitmap_header_ex(s, &bmp->exBitmapDataHeader))
+			return FALSE;
+	}
 
-	if (Stream_GetRemainingLength(s) < cmd->bitmapDataLength)
+	if (Stream_GetRemainingLength(s) < bmp->bitmapDataLength)
 		return FALSE;
 
-	pos = Stream_GetPosition(s) + cmd->bitmapDataLength;
-	cmd->bitmapData = Stream_Pointer(s);
+	pos = Stream_GetPosition(s) + bmp->bitmapDataLength;
+	bmp->bitmapData = Stream_Pointer(s);
 	Stream_SetPosition(s, pos);
-	*length = 20 + cmd->bitmapDataLength;
+
+	return TRUE;
+}
+
+static BOOL update_recv_surfcmd_surface_bits(rdpUpdate* update, wStream* s)
+{
+	SURFACE_BITS_COMMAND* cmd = &update->surface_bits_command;
+
+	if (Stream_GetRemainingLength(s) < 20)
+		return FALSE;
```

Comment:
I guess here the check should be either < 16 or < 16+24

Reply:
right, `8`it is...